File: sample/sample/AdInlineResponsiveViewController.m
//
//  AdInlineResponsiveViewController.m
//  sample
//
//  Created by Vladas Drejeris on 04/05/15.
//  Copyright (c) 2015 Adform. All rights reserved.
//

#import "AdInlineResponsiveViewController.h"
#import <AdformAdvertising/AdformAdvertising.h>

@interface AdInlineResponsiveViewController () <AFAdInlineDelegate>

@property (nonatomic, weak) IBOutlet UIScrollView *scrollView;
@property (nonatomic, weak) IBOutlet UILabel *bottomLabel;
@property (nonatomic, weak) IBOutlet UILabel *topLabel;
@property (nonatomic, weak) IBOutlet NSLayoutConstraint *verticalSpacing;

@end

@implementation AdInlineResponsiveViewController

- (void)viewDidLoad {
    
    // Ad will be loaded in superclass.
    [super viewDidLoad];
    
    // We also need to update constraints to match labels width to screen size.
    [self updateConstraints];
    
    //Setup ad delegate.
    self.adInline.delegate = self;
    
    // Move ad inline to scrollview.
    [self.scrollView addSubview:self.adInline];
    
    // Set adInline view constraints.
    [self setAdInlineConstraints];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSInteger )masterTag {
    
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        return 4022668;
    } else {
        return 4030444;
    }
}

- (void)updateConstraints {
    
    [self.view addConstraint:[NSLayoutConstraint constraintWithItem:self.topLabel
                                                          attribute:NSLayoutAttributeWidth
                                                          relatedBy:NSLayoutRelationLessThanOrEqual
                                                             toItem:self.view
                                                          attribute:NSLayoutAttributeWidth
                                                         multiplier:1.0
                                                           constant:-16]];
    [self.view addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLabel
                                                          attribute:NSLayoutAttributeWidth
                                                          relatedBy:NSLayoutRelationLessThanOrEqual
                                                             toItem:self.view
                                                          attribute:NSLayoutAttributeWidth
                                                         multiplier:1.0
                                                           constant:-16]];
}

- (void)setAdInlineConstraints {
    
    self.adInline.translatesAutoresizingMaskIntoConstraints = NO;
    [self.scrollView addConstraint:[NSLayoutConstraint constraintWithItem:self.adInline
                                                                attribute:NSLayoutAttributeTop
                                                                relatedBy:NSLayoutRelationEqual
                                                                   toItem:self.topLabel
                                                                attribute:NSLayoutAttributeBottom
                                                               multiplier:1.0
                                                                 constant:-8]];
    [self.scrollView addConstraint:[NSLayoutConstraint constraintWithItem:self.adInline
                                                                attribute:NSLayoutAttributeCenterX
                                                                relatedBy:NSLayoutRelationEqual
                                                                   toItem:self.scrollView
                                                                attribute:NSLayoutAttributeCenterX
                                                               multiplier:1.0
                                                                 constant:0]];
}

#pragma mark - AFAdInlineDelegate

- (void)adInlineWillShow:(AFAdInline *)adInline {
    
    // When ad is being displayed we should update label constraints to push it down and make place for the ad.
    // This transition is animated.
    [UIView animateWithDuration:kAdViewAnimationDuration
                     animations:^{
                         [self.verticalSpacing setConstant:(self.verticalSpacing.constant + adInline.adSize.height)];
                         [self.scrollView layoutIfNeeded];
                     }];
}

- (void)adInlineWillHide:(AFAdInline *)adInline {
    
    // When ad is being hidden we should update label constraints and put it back to its initial place.
    // This transition is animated.
    [UIView animateWithDuration:kAdViewAnimationDuration
                     animations:^{
                         [self.verticalSpacing setConstant:(self.verticalSpacing.constant - adInline.adSize.height)];
                         [self.scrollView layoutIfNeeded];
                     }];
}

@end

File: sample/sample/Placements/AdInlineResponsiveViewController.swift
//
//  AdInlineResponsiveViewController.swift
//  sample
//
//  Created by Laurynas Letkauskas on 2021-02-26.
//

import UIKit
import AdformAdvertising

class AdInlineResponsiveViewController: AdInlineViewController {
    
    @IBOutlet private weak var scrollView: UIScrollView!
    @IBOutlet private weak var bottomLabel: UILabel!
    @IBOutlet private weak var topLabel: UILabel!
    @IBOutlet private weak var verticalSpacing: NSLayoutConstraint!
    
    override func viewDidLoad() {
        
        // Ad will be loaded in superclass.
        super.viewDidLoad()
        
        // We also need to update constraints to match labels width to screen size.
        updateConstraints()
        
        //Setup ad delegate.
        adInline.delegate = self
        
        // Move ad inline to scrollview.
        scrollView.addSubview(adInline)
        
        // Set adInline view constraints.
        setAdInlineConstraints()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func masterTag() -> Int {
        987051
    }
    
    func updateConstraints() {
        guard let topLabel = topLabel,
              let bottomLabel = bottomLabel else {
            return
        }
        
        view.addConstraint(
            NSLayoutConstraint(
                item: topLabel,
                attribute: .width,
                relatedBy: .lessThanOrEqual,
                toItem: view,
                attribute: .width,
                multiplier: 1.0,
                constant: -16)
        )
        view.addConstraint(
            NSLayoutConstraint(
                item: bottomLabel,
                attribute: .width,
                relatedBy: .lessThanOrEqual,
                toItem: view,
                attribute: .width,
                multiplier: 1.0,
                constant: -16)
        )
    }
    
    func setAdInlineConstraints() {
        guard let adInline = adInline else {
            return
        }
        
        adInline.translatesAutoresizingMaskIntoConstraints = false
        scrollView.addConstraint(
            NSLayoutConstraint(
                item: adInline,
                attribute: .top,
                relatedBy: .equal,
                toItem: topLabel,
                attribute: .bottom,
                multiplier: 1.0,
                constant: -8))
        scrollView.addConstraint(
            NSLayoutConstraint(
                item: adInline,
                attribute: .centerX,
                relatedBy: .equal,
                toItem: scrollView,
                attribute: .centerX,
                multiplier: 1.0,
                constant: 0))
    }
}

extension AdInlineResponsiveViewController: AFAdInlineDelegate {
    func adInlineWillShow(_ adInline: AFAdInline) {
        // When ad is being displayed we should update label constraints to push it down and make place for the ad.
        // This transition is animated.
        UIView.animate(
            withDuration: TimeInterval(AFAdViewAnimationDuration),
            animations: { [self] in
                verticalSpacing.constant = (verticalSpacing.constant + adInline.adSize.height)
                scrollView.layoutIfNeeded()
            })
    }
    
    func adInlineWillHide(_ adInline: AFAdInline) {
        // When ad is being hidden we should update label constraints and put it back to its initial place.
        // This transition is animated.
        UIView.animate(
            withDuration: TimeInterval(AFAdViewAnimationDuration),
            animations: { [self] in
                verticalSpacing.constant = (verticalSpacing.constant - adInline.adSize.height)
                scrollView.layoutIfNeeded()
            })
    }
}
